### PR TITLE
Elastic split

### PR DESCRIPTION
**Pull request by ctdk**
And the nsq_probe_metrics_to_elasticsearch half of the split indexes.

### CODE REVIEW

File: eventdef/eventdef.go [hunk 1]
```diff
@@ -57,12 +86,77 @@ func Save(e *schema.ProbeEvent) error {
 	if err := e.Validate(); err != nil {
 		return err
 	}
+	
+	t := time.Now()
+	y, m, d := t.Date()
+	idxName := fmt.Sprintf("events-%d-%02d-%02d", y, m, d)
+	if err := checkIdx(idxName, e); err != nil {
```
Comment (woodsaj) [inline eventdef/eventdef.go]: You dont need to do this.. Use a index template in Elasticsearch and the index will be automatically created when first written to.
https://www.elastic.co/guide/en/elasticsearch/reference/1.6/indices-templates.html

**Comment by woodsaj** [comment]
The bulk indexer does not track errors well.  If 1 document from a batch fails, we currently have no way to identify which document it was and either NACK it or re-queue it.

The problem is not with the response from Elastic, but with the way elastigo processes the response.  We can address this issue by providing our own bulkIndexer.Send method instead of the default:
https://github.com/mattbaird/elastigo/blob/master/lib/corebulk.go#L335

**Comment by Dieterbe** [comment]
Why does this add init scripts for raintank-metric, a service we no longer use and which has nothing to do with es bulk loading?

**Comment by ctdk** [comment]
Those were left over from some work on packaging from a long time ago and got caught up in a commit. I did notice, but hadn't removed them from the PR yet. Just did so.

**Comment by ctdk** [comment]
Incidentally, the rest of this PR is in flux right now anyway - I've removed the checkIdx function (to be replaced by templates, which is a lot more straightforward), and am writing a custom bulk sender.

**Comment by Dieterbe** [comment]
@ctdk let us know when this is ready for another round of review

**Comment by ctdk** [comment]
@Dieterbe I just need to find a way to test this case: https://github.com/raintank/raintank-metric/pull/35/files#diff-d2f0dff8d8d716c6b4a29c985105686eR154

As of now nsq probe recovers and requeues just fine if elasticsearch itself is down. However, I haven't yet figured out how to test the case where elasticsearch is running, but one or more items fail to save for some reason. Also since I'm not sure what elasticsearch sends back in that case, I haven't been able to write the recovery code properly anyway.

If you have any ideas there I'd love to hear them.

**Comment by Dieterbe** [comment]
so basically you need a way to have ES reliable send failures back so you can see and test with those responses? not sure what's the easiest way to "hurt" ES, maybe fill up its data volume or set a replication factor that can't be met. or maybe they have docs on this? they should

**Comment by ctdk** [comment]
I've tried filling up ES's partition and that didn't work (but, I'm happy to say, it also didn't break entirely - nsq_probe_events_to_elasticsearch just waited until it started working again to start submitting events), nor did trying to force having duplicate ids. Trying to find other ways to force the issue.

**Comment by ctdk** [comment]
Found a way - I changed the elastigo library to "create" instead of "index", forcing an error when an id is duplicated. This is ready to be reviewed again. @woodsaj @Dieterbe

File: eventdef/eventdef.go [hunk 2]
```diff
@@ -57,12 +95,111 @@ func Save(e *schema.ProbeEvent) error {
 	if err := e.Validate(); err != nil {
 		return err
 	}
+	
+	t := time.Now()
+	y, m, d := t.Date()
+	idxName := fmt.Sprintf("events-%d-%02d-%02d", y, m, d)
```
Comment (woodsaj) [inline eventdef/eventdef.go]: the indexName needs to be based on the event timestamp, not on when it gets written.

**Comment by woodsaj** [comment]
This code is hard to follow and looks very fragile.  Can you 
a) add comments to explain why and what you are doing.
b) add unit tests (yes, i am a hypocrite, but they are needed)
c) The logging is too verbose for production. So either remove it or only print it if the app is running in debug mode.

**Comment by Dieterbe** [comment]
k

**Comment by ctdk** [comment]
@woodsaj Addressing those now. a) and c) won't take too long. b) will require a little more thought.

**Comment by ctdk** [comment]
@woodsaj a) and c) have been addressed. Working on b).

**Comment by woodsaj** [comment]
I have merged master into this branch and things are working, but i see lots of issues with the design.

1. number of handlers is hard coded at 80
2. each handler blocks until the event has been sent to Elasticsearch.  This means only 80 events will be buffered at a time and will block the entire app until they are saved in elasticsearch, which by default is every 5seconds.  So the maximum performance we can see is 16events/second.  That is not going to be anywhere near enough for production.
3. There is a race condition where the Id can bet added to bSender.Queue, then bulkSend() called before the event has been added to the ES bulkIndexer.  This will result in a handler goroutine blocking indefinitely.

**Comment by woodsaj** [comment]
just pushed a major refactor of this to address the issues raised in my earlier comment.

Some things to note when deploying this.  The nsq messages are marked as "inflight" until they are acked which doesnt happen until after the event is committed to ES.  As such inFlight needs to be at least as large as we want to have messages in the batch.  ES suggests that 5-15Mb is a good size for bulk requests. https://www.elastic.co/guide/en/elasticsearch/guide/current/bulk.html
